src/v/config/property.h

```diff
@@ -123,10 +131,180 @@ class property : public base_property {
+    return "string";
+}
+
+template<typename T>
+concept has_type_name = requires(T x) {
+    x.type_name();
+};
+
+template<typename T>
+requires has_type_name<T> std::string_view property_type_name(T*) {
+    return T::type_name();
+}
+
+template<typename T>
+struct is_optional : std::false_type {};
+
+template<typename T>
+struct is_optional<const std::optional<T>> : std::true_type {};
+
+template<typename T>
+requires is_optional<T>::value std::string_view property_type_name(T*) {
+    typename T::value_type const* ptr = nullptr;
+    return property_type_name(ptr);
+}
+
+template<typename T>
+concept is_collection = requires(T x) {
+    typename T::value_type;
+    {x.size()};
+};
+
+template<typename T>
+requires is_collection<T> std::string_view property_type_name(T*) {
+    typename T::value_type const* ptr = nullptr;
+    return property_type_name(ptr);
+}
+
+template<typename T>
+std::string_view property<T>::type_name() const {
+    // Go via a non-member function so that specialized implementations
+    // can use concepts without all the same concepts having to apply
+    // to the type T in the class definition.
+    return property_type_name(&_value);
+}
+
+inline std::optional<std::string_view>
+property_units_name(const std::chrono::milliseconds*) {
+    return "ms";
+}
+
+inline std::optional<std::string_view>
+property_units_name(const std::chrono::seconds*) {
+    return "s";
+}
+
+// Default units is null -- e.g. any non-numeric setting
+template<typename T>
+inline std::optional<std::string_view> property_units_name(T*) {
+    return std::nullopt;
+}
+
+template<typename T>
+requires config::is_optional<T>::value std::optional<std::string_view>
+property_units_name(T*) {
+    typename T::value_type const* ptr = nullptr;
+    return property_units_name(ptr);
+}
+
+template<typename T>
+std::optional<std::string_view> property<T>::units_name() const {
+    return property_units_name(&_value);
+}
+
+template<typename T>
+inline bool property_is_nullable() {
+    return false;
+}
+
+template<typename T>
+requires is_optional<T>::value inline bool property_is_nullable() {
+    return true;
+}
+
+template<typename T>
+bool property<T>::is_nullable() const {
+    return property_is_nullable<const T>();
+}
+
+template<typename T>
+inline bool property_is_array() {
+    return false;
+}
+
+template<typename T>
+requires is_collection<T>
+inline bool property_is_array() { return true; }
+
+template<typename T>
+bool property<T>::is_array() const {
+    return property_is_array<T>();
+}
+
 /*
- * Same as property<std::vector<T>> but will also decode a single T. This can be
- * useful for dealing with backwards compatibility or creating easier yaml
- * schemas that have simplified special cases.
+ * Same as property<std::vector<T>> but will also decode a single T. This
+ * can be useful for dealing with backwards compatibility or creating easier
+ * yaml schemas that have simplified special cases.
```

Comment:
Looks like you forgot a `}` at some point!

Reply:
Fixed.